### DISCUSSION TITLE
How do the Tab and arrow keys work under the hood when navigating MenuComplete completions?

### DISCUSSION THREAD

**Discussion by matteocoder** [lzybkr/PSReadLine#3835] [Q&A] [answered]
First of all, thank you all for working on this amazing piece of software. I've recently switched to PowerShell, and creating new functionality using PSReadLine's functions was easier than I thought. I've also been paying particular attention to keyboard ergonomics, and how to adjust key bindings to make using a computer more comfortable for my fingers. 

When I invoke Menu completion with ctrl+spacebar, it is possible to navigate the suggestions using either the tab key or the arrow keys. I have been trying to remap ctrl+h,j,k,l to the PSReadLine functions which the arrow keys use under the hood, so that I could navigate the suggestions using familiar vim motions. Unfortunately, those mappings do not seem to work in menu completion; For instance, if I map ctrl+l to the right arrow key, and press ctrl+l to navigate to the right in the menu, I actually move one character to the right in the text buffer.

Thus, here are some questions I have:
- how does the invocation of Menu completion alter the behaviour of the tab and arrow keys?
- Are there any externally available functions or methods which the user can bind to arbitrary key chords to mimic this particular behaviour of the tab and arrow keys?

For reference, [this StackOverflow user](https://stackoverflow.com/questions/53886446/hjkl-for-navigation-on-powershell-psreadlines-menucomplete-function) presents the same issue, but the answer is non-satisfactory.

Note that the version of PSReadline I'm using is 2.3.4

**Comment by daxian-dbw** [comment, marked as answer]
Key pressing in the menu completion context doesn't go through the key-binding handling.
While in the menu completion, it reads key directly through `Console.ReadKey()`, and then acts based on the key.

https://github.com/PowerShell/PSReadLine/blob/ac69010fec57df06827aebb711d243706ec24e9b/PSReadLine/Completion.cs#L939-L946

So, those actions cannot be bound in a custom way.

**Reply by matteocoder** [reply]
Thank you for your answer, @daxian-dbw. In this case, I imagine that the only way for me to customise the keys which move through the completions is to use an external program like PowerToys or AutoHotkey and emulate the arrow keys / tab key.
